sdlib/d/gc/page.d

```diff
@@ -784,21 +825,63 @@ private:
 				if (ec.isLarge()) {
 					if (w != gcCycle) {
 						// We have not marked this extent this cycle.
+						// check for a finalizer
+						auto f = e.finalizer;
+						if (f !is null) {
+							__sd_run_finalizer(e.address, e.usedCapacity, f);
+						}
+
 						deadExtents.insert(e);
 					}
 
 					continue;
 				}
 
-				// If the cycle do not match, all the elements are dead.
-				if ((w & 0xff) != gcCycle) {
+				// need to know if anything has metadata
+				auto hasMeta = e.slabMetadataFlags.rawNimbleAtomic(0);
+
+				// If the cycle do not match, all the elements are dead. If no
+				// metadata exists, then we can safely clear the whole thing.
+				if (hasMeta == 0 && ((w & 0xff) != gcCycle)) {
 					deadExtents.insert(e);
 					continue;
 				}
 
 				auto oldOccupancy = e.slabData.rawContent[0];
 				auto newOccupancy = oldOccupancy & (w >> 8);
 
+				// need to check dying blocks for finalizers.
+				hasMeta &= (oldOccupancy ^ newOccupancy);
+				if (hasMeta) {
+					import d.gc.slab;
+					auto ssize = binInfos[ec.sizeClass].slotSize;
+					// All set bits in hasMeta have metadata.
+					auto baseidx = i * 64;
+					while (hasMeta != 0) {
+						uint bit = countTrailingZeros(hasMeta);
+						// NOTE: this copies techniques/code from
+						// SlabAllocInfo, but that code starts with a
+						// pointer and gives us info we already have.
+						// So we don't want to do that work again.
```

Comment:
It might be worth thinking about refactorings here.

Reply:
You mean a refactor in slab.d? I need this specific calculation abstracted out I guess.

Reply:
Removed.